### ISSUE TITLE
Watch_broadcasts should warn or not work if registry not properly configured

### ISSUE THREAD

**Issue by timothyfroehlich** [iotile/coretools#673]
There is a registry setting that is required for watch_broadcasts to work. Without that setting, it will simply find no results. Either watch_broadcasts should refuse to work or it should warn that this setting is required to get results.

**Comment by timburke** [comment]
I agree this would be a great feature since it always trips people up.  It's going to take a bit of work to make it happen though.  The primary reason why it's an issue is also why it's a bit difficult to fix correctly.

The `HardwareManager` class that you are calling `enable_broadcasts` on does not actually know if anything needs to be done in order for the underlying transport plugin to actually receive broadcasts.  It just passes on the message to enable the broadcast interface, whatever that means.

Of the transport plugins that we have implemented (`websockets`, `bled112`, `nativeble`, `mqtt`, `virtual`, `emulated`), only `bled112` requires a registry setting for it to properly receive broadcast packets and only if you want to receive those broadcasts from a POD using the v1 bluetooth advertisement format.  

There's not enough context in the `enable_broadcasts` method call for `HardwareManager` to determine whether you should have had `bled112:active-scan = True`.  Maybe you wanted to only receive `v2` broadcast packets that do not require that setting and you want the BLED112 dongle to 100% receive only so `active-scan` is actively disabled.  Maybe you are using a transport plugin that always can receive broadcast packets (like all of the non-ble options).

Once thing we could do is handle open_interface call on the broadcast interface in the `bled112` plugin and log a warning that you will not receive v1 broadcasts if `active-scan` is disabled.  You would still need to know to display the log and look there though.

**Comment by mattrunchey** [comment]
I just did a quick test, we could implement something as simple as a check in `watch_broadcasts` based on the transport you're running the hwmanager on:

```
        if self.transport == 'bled112':
            reg = ComponentRegistry()
            self.logger.info(reg.get_config('bled112:active-scan'))
            if reg.get_config('bled112:active-scan') == 'false':
                self.logger.warning("Warning: active scan not set, you may not see results")
                time.sleep(2)
```

However, this wouldn't solve if you're connected to a ws `iotile-gateway` that was invoked without `active-scan` set.

**Comment by timburke** [comment]
Yeah this could be a good useability workaround.

**Comment by vvynohradov** [comment]
We should also check if `bled112:active-scan` key exists
![image](https://user-images.githubusercontent.com/41442348/52958507-dd2adf80-339c-11e9-8a0f-c956440a4f7c.png)

**Comment by vvynohradov** [comment]
And, I think that we should print this warning inside `watch_scan` ui, because it is not really understandable if the warning occured, without verbose logging(-v flags), shouldn't we? Something like this
![image](https://user-images.githubusercontent.com/41442348/52962076-0bacb880-33a5-11e9-9e10-5d6d61f892a7.png)

**Comment by timburke** [comment]
> We should also check if `bled112:active-scan` key exists
> ![image](https://user-images.githubusercontent.com/41442348/52958507-dd2adf80-339c-11e9-8a0f-c956440a4f7c.png)

This warning is because the config value is accessed directly via the `ComponentRegistry` rather than by using the `ConfigManager` [1].  The config manager handles providing a default value for config variables in case they are not set.  It uses the `ComponentRegistry` as the underlying kv store but adds an additional layer of type safety and default values:

`ComponentRegistry.get_config(name)` is equivalent to `ConfigManager.get(name)` except that the result is a native python type, not string and default values are returned if the variable is not overridden.  The defaults are set by the package that registered the config variable, in this case here:

https://github.com/iotile/coretools/blob/master/transport_plugins/bled112/iotile_transport_bled112/config_variables.py#L9

In terms of showing the error in the watch ui itself.  I agree that is a good idea.  I would rephrase the warning though as:
"Warning: active scan not set, you will not see v1 format broadcasts"

since new POD firmwares use a different broadcast format that does not require active scanning.


[1] https://github.com/iotile/coretools/blob/master/iotilecore/iotile/core/dev/config.py#L120

**Comment by vvynohradov** [comment]
Thank you for clarifying, I think it is better to wrap it up wtih try/except to prevent this exception warning.

But also, why we can't check tile version, before showing this warning? And then we will know format of broadcast message, and won't show warning message in case of different format

**Comment by timburke** [comment]
> Thank you for clarifying, I think it is better to wrap it up wtih try/except to prevent this exception warning.
> 
> But also, why we can't check tile version, before showing this warning? And then we will know format of broadcast message, and won't show warning message in case of different format

`watch_broadcasts` never makes a connection to a POD so it is unable to send it any RPCs.  Put another way, the `broadcast` data sent from PODs is undirected pub/sub type communications that are sent like bluetooth beacons.  When `watch_broadcasts` is used, the computer running that command is just deciding to listen for these beacons.  The goal of using watch_broadcasts instead of, for example, connecting to a POD and then using `enable_streaming` to get readings from that POD is to be able to listen to many PODs simultaneously without needing to connect to them.

There's no 1-1 connection with any given POD that could be used to query its configuration.

**Comment by timburke** [comment]
> I think it is better to wrap it up wtih try/except to prevent this exception warning.

I would recommend just using the `ConfigManager` instead of a try/except.  `ConfigManager` exists to prevent just this exception when accessing a config variable whose default value has not been overridden.  In general, config variables should not be accessed directly through `ComponentRegistry`.  It's just an implementation detail that they happen to be stored in the same place as registered components.

**Comment by vvynohradov** [comment]
> `watch_broadcasts` never makes a connection to a POD so it is unable to send it any RPCs. Put another way, the `broadcast` data sent from PODs is undirected pub/sub type communications that are sent like bluetooth beacons. When `watch_broadcasts` is used, the computer running that command is just deciding to listen for these beacons. The goal of using watch_broadcasts instead of, for example, connecting to a POD and then using `enable_streaming` to get readings from that POD is to be able to listen to many PODs simultaneously without needing to connect to them.
> 
> There's no 1-1 connection with any given POD that could be used to query its configuration.

Yes, thank you, that make sense

> I would recommend just using the `ConfigManager` instead of a try/except. `ConfigManager` exists to prevent just this exception when accessing a config variable whose default value has not been overridden. In general, config variables should not be accessed directly through `ComponentRegistry`. It's just an implementation detail that they happen to be stored in the same place as registered components.

Ok, right, it will be more logical architecture
